File: packages/connectors/src/google/tasks.ts
/**
 * Google Tasks connector — task lists, CRUD, completion.
 */

import type { Task, TaskList, CreateTaskParams, UpdateTaskParams, Provider } from '../types.js';
import type { GoogleClient } from './api-client.js';

const TASKS = '/tasks/v1';

interface GTask { id: string; title: string; notes?: string; due?: string; status: string; completed?: string }

export class GoogleTasksConnector {
  private readonly provider: Provider = 'google';
  constructor(private client: GoogleClient, private accountId: string) {}

  async listTaskLists(): Promise<TaskList[]> {
    const data = await this.client.get<{ items?: { id: string; title: string }[] }>(
      `${TASKS}/users/@me/lists`, undefined, { accountId: this.accountId },
    );
    return (data.items ?? []).map((l) => ({ id: l.id, accountId: this.accountId, provider: this.provider, name: l.title }));
  }

  async listTasks(listId?: string): Promise<Task[]> {
    const lid = listId ?? '@default';
    const data = await this.client.get<{ items?: GTask[] }>(
      `${TASKS}/lists/${lid}/tasks`, { showCompleted: 'false', showHidden: 'false' }, { accountId: this.accountId },
    );
    return (data.items ?? []).map((t) => this.parse(t, lid));
  }

  async createTask(params: CreateTaskParams): Promise<Task> {
    const lid = params.listId ?? '@default';
    const body: Record<string, unknown> = { title: params.title, notes: params.notes };
    if (params.dueDate) body.due = params.dueDate.toISOString();
    const data = await this.client.post<GTask>(`${TASKS}/lists/${lid}/tasks`, body, { accountId: params.accountId ?? this.accountId });
    return this.parse(data, lid);
  }

  async updateTask(params: UpdateTaskParams): Promise<Task> {
    const lid = params.listId ?? '@default';
    const body: Record<string, unknown> = {};
    if (params.title !== undefined) body.title = params.title;
    if (params.notes !== undefined) body.notes = params.notes;
    if (params.dueDate !== undefined) body.due = params.dueDate.toISOString();
    const data = await this.client.patch<GTask>(`${TASKS}/lists/${lid}/tasks/${params.taskId}`, body, { accountId: params.accountId ?? this.accountId });
    return this.parse(data, lid);
  }

  async completeTask(taskId: string, listId?: string): Promise<Task> {
    const lid = listId ?? '@default';
    const data = await this.client.patch<GTask>(`${TASKS}/lists/${lid}/tasks/${taskId}`, { status: 'completed' }, { accountId: this.accountId });
    return this.parse(data, lid);
  }

  async deleteTask(taskId: string, listId?: string): Promise<void> {
    await this.client.delete(`${TASKS}/lists/${listId ?? '@default'}/tasks/${taskId}`, { accountId: this.accountId });
  }

  private parse(task: GTask, listId: string): Task {
    return {
      id: task.id, accountId: this.accountId, provider: this.provider,
      title: task.title, notes: task.notes,
      dueDate: task.due ? new Date(task.due) : undefined,
      isCompleted: task.status === 'completed',
      completedAt: task.completed ? new Date(task.completed) : undefined,
      listId,
    };
  }
}

File: services/api/app/google/tasks.py
"""Direct Google Tasks API client using stored OAuth tokens."""

import httpx
from .oauth import get_valid_token

TASKS_API = "https://tasks.googleapis.com/tasks/v1"


def _headers() -> dict:
    token = get_valid_token("tasks")
    if not token:
        raise ConnectionError("Google Tasks not connected. Complete OAuth setup first.")
    return {"Authorization": f"Bearer {token}"}


def get_task_lists() -> list[dict]:
    """Get all task lists."""
    with httpx.Client(timeout=15) as client:
        resp = client.get(f"{TASKS_API}/users/@me/lists", headers=_headers())
        resp.raise_for_status()
    return [{"id": t["id"], "title": t["title"]} for t in resp.json().get("items", [])]


def get_pending_tasks(tasklist_id: str = "@default") -> list[dict]:
    """Get pending (not completed) tasks from a list."""
    with httpx.Client(timeout=15) as client:
        resp = client.get(
            f"{TASKS_API}/lists/{tasklist_id}/tasks",
            headers=_headers(),
            params={"showCompleted": "false", "showHidden": "false"},
        )
        resp.raise_for_status()

    return [
        {
            "id": t["id"],
            "title": t.get("title", ""),
            "notes": t.get("notes", ""),
            "due": t.get("due"),
            "status": t.get("status"),
        }
        for t in resp.json().get("items", [])
        if t.get("title")
    ]


def create_task(title: str, notes: str = "", tasklist_id: str = "@default") -> dict:
    """Create a new task."""
    with httpx.Client(timeout=15) as client:
        resp = client.post(
            f"{TASKS_API}/lists/{tasklist_id}/tasks",
            headers={**_headers(), "Content-Type": "application/json"},
            json={"title": title, "notes": notes},
        )
        resp.raise_for_status()

    data = resp.json()
    return {"id": data["id"], "title": data["title"], "status": data.get("status")}


def update_task(task_id: str, updates: dict, tasklist_id: str = "@default", email: str | None = None) -> dict:
    """Update a task. updates can include: title, notes, due, status."""
    token = get_valid_token("tasks", email)
    if not token:
        raise ConnectionError("Google Tasks not connected.")
    with httpx.Client(timeout=15) as client:
        resp = client.patch(
            f"{TASKS_API}/lists/{tasklist_id}/tasks/{task_id}",
            headers={"Authorization": f"Bearer {token}", "Content-Type": "application/json"},
            json=updates,
        )
        resp.raise_for_status()
    data = resp.json()
    return {"id": data["id"], "title": data.get("title"), "status": data.get("status")}


def complete_task(task_id: str, tasklist_id: str = "@default", email: str | None = None) -> dict:
    """Mark a task as completed."""
    return update_task(task_id, {"status": "completed"}, tasklist_id, email)


def delete_task(task_id: str, tasklist_id: str = "@default", email: str | None = None) -> dict:
    """Delete a task."""
    token = get_valid_token("tasks", email)
    if not token:
        raise ConnectionError("Google Tasks not connected.")
    with httpx.Client(timeout=15) as client:
        resp = client.delete(
            f"{TASKS_API}/lists/{tasklist_id}/tasks/{task_id}",
            headers={"Authorization": f"Bearer {token}"},
        )
        resp.raise_for_status()
    return {"id": task_id, "deleted": True}
